lib/phoenix/channel/server.ex

```diff
@@ -10,55 +10,91 @@ defmodule Phoenix.Channel.Server do
 
   ## Transport API
 
+  @doc """
+  Starts a channel server.
+  """
+  def start_link(socket, payload, pid, ref) do
+    GenServer.start_link(__MODULE__, {socket, payload, pid, ref})
+  end
+
   @doc """
   Joins the channel in socket with authentication payload.
   """
-  @spec join(Socket.t, map) :: {:ok, map, pid} | {:error, map}
-  def join(socket, auth_payload) do
-    Phoenix.Endpoint.instrument socket, :phoenix_channel_join,
-      %{params: auth_payload, socket: socket}, fn ->
+  @spec join(Socket.t, module, Message.t, keyword) :: {:ok, map, pid, Socket.t} | {:error, map}
+  def join(socket, channel, message, opts) do
+    %{topic: topic, payload: payload, ref: join_ref} = message
+    assigns = Keyword.get(opts, :assigns, %{})
+
+    socket =
+      %Socket{
+        socket
+        | topic: topic,
+          channel: channel,
+          join_ref: join_ref,
+          assigns: Map.merge(socket.assigns, assigns),
+          private: channel.__socket__(:private)
+      }
+
+    instrument = %{params: payload, socket: socket}
+
+    Phoenix.Endpoint.instrument socket, :phoenix_channel_join, instrument, fn ->
       ref = make_ref()
 
-      case GenServer.start_link(__MODULE__, {socket, auth_payload, self(), ref}) do
+      case Supervisor.start_child(socket.handler, [socket, payload, self(), ref]) do
```

Comment:
Do I get this right that this PR also changes the process structure to introduce a dedicated supervisor for each socket connection?

Reply:
Yes!

Reply:
Cool! So I guess that this might also fix #2466 then.